Replace solve by solveset in inequalities

hargup opened the pull request:
fixes: #9567

hargup committed:
Replace solve by solveset in inequalities

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

hargup commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -301,5 +301,9 @@ def test_issue_8545():
+
+
+@XFAIL
+def test_issue_8974_fails():
```
Not sure why this fails.

hargup commented:
This also fixes #9567 after https://github.com/sympy/sympy/issues/10023 is fixed.

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -396,7 +396,17 @@ def solve_univariate_inequality(expr, gen, relational=True):
+    from sympy.solvers.solvers import denoms
+    from sympy.solvers.solveset import solveset
+
+    def solve_replacement(f_x, x):
```
You could just capture `solve` as `sympy_solve=solve` and then `def solve` instead of `def solve_replacement` using `sympy_solve` internally so no other changes would be necessary in the file.

asmeurer replied:
The name `solve_replacement` makes sense in the context of this PR, but will it make sense a year from now in this file when people come across it?

hargup replied:
> The name `solve_replacement` makes sense in the context of this PR, but will it make sense a year from now in this file when people come across it?

Probably it won't, I like @smichr 's suggestion to keep the name of this function as `solve` so that we won't have to change the actual code. This function which emulates the output api of `solve` using `solveset` is of more general use than just this file hence I'm moving it to `solveset`. We will have two solves:
```python
sympy.solvers.solvers.solve # The actual old solve
sympy.solvers.solveset.solve # emulation of old solve by solveset
```

hargup committed:
Merge remote-tracking branch 'sympy/master' into ineq_solveset

* sympy/master: (30 commits)
  docs changes in rv.py
  Union(Interval(-oo, oo), FiniteSet(1)) returns Interval(-oo, oo)
  Import pi.
  9954: expressions can change to bool when solving inequality
  Move test to  the end of core/test_assumptions
  edit minsolve_linear_system docstring
  modifications of solve_linear
  Add test for Mod -> mod in numpy lambdify.
  use single element rather than single element list
  factor free_symbol calculation out of comprehension
  ...

hargup committed:
Emulate solve using solveset

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

hargup committed:
Replace solve by solveset

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

hargup commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -308,5 +308,9 @@ def test_issue_8545():
+
+
+@XFAIL
+def test_issue_8974_fails():
```
I'm not sure why this fails.

hargup committed:
Add docstring for solvers.solveset.solve

Signed-off-by: Harsh Gupta <gupta.harsh96@gmail.com>

smichr committed:
watch for no solutions when building intervals

hargup committed:
Merge pull request #13 from smichr/ineq

watch for no solutions when building intervals

smichr committed:
correct slow_general_univariate test

The old test was wrong as was confirmed graphically. The solution to the
new expression comes back as a non-RootOf expression which has trouble
computing at low precision. It will be better to send such roots back
as RootOf instances, but that can be done in a subsequent commit.

For the record, here is the evaluation of the interval at different
precisions:

>>> i=solve(sqrt(x) + 1/root(x, 3) - 2 > 0).as_set()
>>> for n in range(3, 6):
...  print n, i.n(n)
...
3 (0, 0.375) U (1.0, +inf)
4 (0, 0.3727) U (1.0, +inf)
5 (0, 0.37284) U (1.0, +inf)

The true answer can be obtained numerically from nsolve:

>>> nsolve(sqrt(x) + 1/root(x, 3) - 2, .37)
mpf('0.37284373759817781')

hargup committed:
Merge pull request #14 from smichr/ineq

correct slow_general_univariate test

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -284,6 +279,11 @@ def test_solve_univariate_inequality():
 
 
+def test_solve_univariate_inequality_errors():
+    raises(NotImplementedError, lambda: isolve(sin(x) < S.Half, x))
```
Looks like there is a duplicate test in `test_solvers/test_inequalities` -- the last line of that test function.

hargup commented:
Please feel free to take this up.

kshitij10496 commented:
@hargup I would like to continue on this.

hargup commented:
> @hargup I would like to continue on this. 

Go ahead, you can add a remote to my repo and pull these changes.

kshitij10496 commented:
@hargup 
Thanks.
I did that.
